### PR TITLE
feat(read): perform reading from storage in parallel for faster reads

### PR DESCRIPTION
**Pull request by lionel1704**
Note that this PR also contains some changes to allow parallel writes, but this is not supported by sn_client at the moment and hence it is disabled.

### CODE REVIEW

File: src/self_encryptor.rs [hunk 1]
```diff
@@ -266,6 +267,7 @@ where
             }
         }
 
+        // let mut network_storage_futures = vec![];
```
Comment (maqi) [inline src/self_encryptor.rs:270]: forget to remove?

File: src/self_encryptor.rs [hunk 2]
```diff
@@ -281,16 +283,21 @@ where
...
 
-                self.storage.put(name.to_vec(), content).await?;
-
                 new_map[i].hash = name.to_vec();
+                self.storage.put(name.to_vec(), content).await?;
+                // let mut storage = self.storage.clone();
+                // network_storage_futures.push(async move { storage.put(name.to_vec(), content).await });
             }
         }
+        // let results = join_all(network_storage_futures.into_iter()).await;
```
Comment (maqi) [inline src/self_encryptor.rs:292]: removed?

File: src/self_encryptor.rs [hunk 3]
```diff
@@ -343,11 +352,15 @@ where
...
-            let vec = decrypt_chunk(&mut *state, i).await?;
-            decrypted_chunks.push((vec, pos));
+            positions.push(get_start_end_positions(state.file_size, i).0);
+            decryption_futures.push(decrypt_chunk(&mut *state, i).await);
         }
     }
+    let decrypted_data = join_all(decryption_futures).await;
+    let mut pos_iter = positions.into_iter();
+    for chunk in decrypted_data {
+        decrypted_chunks.push((chunk?, pos_iter.next().unwrap_or(0)))
```
Comment (maqi) [inline src/self_encryptor.rs:362]: `any` of failed decryption will cause `all left over` decrypted chunk to be dropped ?

File: src/self_encryptor.rs [hunk 4]
```diff
@@ -508,16 +521,23 @@ where
...
-        let chunk_data = decrypt_chunk(&mut *state, i).await?;
-        decrypted_chunks.push((chunk_data, pos));
+        positions.push(get_start_end_positions(state.file_size, i).0);
+        decryption_futures.push(decrypt_chunk(&mut *state, i).await);
+    }
+
+    let chunks = join_all(decryption_futures.into_iter()).await;
+    let mut pos_iter = positions.into_iter();
+    for chunk in chunks {
+        decrypted_chunks.push((chunk.unwrap(), pos_iter.next().unwrap_or(0)))
```
Comment (maqi) [inline src/self_encryptor.rs:540]: use `?` or `unwrap` ?

File: src/sequential/large_encryptor.rs [hunk 5]
```diff
@@ -143,7 +159,7 @@ where
         data = self.fill_chunk_buffer(data, 0).await?;
         data = self.fill_chunk_buffer(data, 1).await?;
 
-        let mut all_things = Vec::new();
+        // let mut storage_futures = Vec::new();
```
Comment (maqi) [inline src/sequential/large_encryptor.rs:162]: remove?

File: src/sequential/large_encryptor.rs [hunk 6]
```diff
@@ -156,9 +172,14 @@ where
                 let mut data_to_encrypt = self.buffer.split_off(MAX_CHUNK_SIZE);
                 mem::swap(&mut self.buffer, &mut data_to_encrypt);
                 let index = self.chunks.len();
-                all_things.push(self.encrypt_chunk(&data_to_encrypt, index).await?);
+                self.encrypt_chunk(&data_to_encrypt, index).await?.await?
+                // storage_futures.push(self.encrypt_chunk(&data_to_encrypt, index).await?);
             }
         }
+        // let results = futures::future::join_all(storage_futures.into_iter()).await;
```
Comment (maqi) [inline src/sequential/large_encryptor.rs:179]: remove?

File: src/sequential/medium_encryptor.rs [hunk 7]
```diff
@@ -18,31 +20,38 @@ pub const MAX: usize = 3 * MAX_CHUNK_SIZE;
...
             let pad_key_iv = utils::get_pad_key_and_iv(index, &chunks);
-            let chunk = storage.get(&chunk.hash).await?;
+            let mut storage = storage.clone();
+            get_futures
+                .push(async move { storage.get(&chunk.hash).await.map(|res| (res, pad_key_iv)) });
+        }
+        let results = join_all(get_futures.into_iter()).await;
+        for chunk in results {
+            let (chunk, pad_key_iv) = chunk?;
             let decrypted_chunk = utils::decrypt_chunk(&chunk, pad_key_iv)?;
```
Comment (maqi) [inline src/sequential/medium_encryptor.rs:55]: can this `decrypt` operation to be moved into the `futures` as well? i.e. the result is `decrypted_chunk` directly?
i.e. the async function does `get` and `decrypt` in one go

File: src/sequential/medium_encryptor.rs [hunk 8]
```diff
@@ -78,7 +87,7 @@ where
             return Ok((DataMap::Chunks(chunks), self.storage));
         }
 
-        let mut all_chunks;
+        // let mut chunk_storage_futures;
```
Comment (maqi) [inline src/sequential/medium_encryptor.rs:90]: remove?

File: src/sequential/medium_encryptor.rs [hunk 9]
```diff
@@ -113,9 +122,15 @@ where
 
                 let hash = self.storage.generate_address(&encrypted_contents).await?;
                 details.hash = hash.to_vec();
-                all_chunks.push(self.storage.put(hash.to_vec(), encrypted_contents).await?);
+                self.storage.put(hash.to_vec(), encrypted_contents).await?
+                // let mut storage = self.storage.clone();
```
Comment (maqi) [inline src/sequential/medium_encryptor.rs:126]: remove?

File: src/test_helpers.rs [hunk 10]
```diff
@@ -40,27 +41,39 @@ impl<'a> Debug for Blob<'a> {
...
+        }
     }
 
-    pub fn has_chunk(&self, name: &[u8]) -> bool {
-        self.entries.iter().any(|entry| entry.name == name)
+    pub async fn has_chunk(&self, name: &[u8]) -> Result<bool, SelfEncryptionError> {
+        Ok(self
+            .entries
+            .read()
+            .map_err(|_| SelfEncryptionError::Poison)?
```
Comment (maqi) [inline src/test_helpers.rs:66]: better at least log the error?

Comment (lionel1704) [reply src/test_helpers.rs:66]: This error happens for test only and there is only one possible error. So I think we should be okay here.

**Comment by lionel1704** [comment]
@maqi I've push a comment addressing your comments. The code that is commented out is left there on purpose. As I've mentioned in the PR description, it meant for the parallel writes which is not yet supported by sn_client. They will be un-commented once that is ready.

File: src/self_encryptor.rs (hunk 4 above)
Comment (dirvine) [reply src/self_encryptor.rs]: Not unwrap of course ;-)

**Review by dirvine — APPROVED** [review]

**Pull request merged**